### PR TITLE
Fix Java code generation for `CALL BY VALUE` arguments

### PR DESCRIPTION
**Pull request by tsh-hashimoto**
https://github.com/opensourcecobol/opensourcecobol4j/issues/785 に関する修正

# 概要

`CALL BY VALUE` で数値引数を渡す際のJavaコード生成を修正する。

従来のコード生成では、C言語固有の型キャスト（`unsigned`, `char`, `short`, `long long` など）がそのままJavaコードとして出力されており、不正なJavaコードが生成されていた。
本PRでは、`CobolDataStorage.primitiveToDataStorage()` を使用して数値をバイト列に変換する方式に変更し、正しいJavaコードが生成されるようにした。

# 変更点

## `cobj/codegen.c`

- `CALL BY VALUE` の数値引数に対するJavaコード生成を修正
  - C言語の型キャスト（`(unsigned char)`, `(short)`, `(int)`, `(long long)` など）の出力を削除
  - SIZE 8の場合: `CobolDataStorage.primitiveToDataStorage((long)(...))` を出力
  - それ以外（SIZE 1, 2, 4）の場合: `CobolDataStorage.primitiveToDataStorage((int)(...))` を出力
- `SIZE AUTO` の場合に出力されていた `(unsigned ` のキャスト（C言語構文）を削除
- `default` ブランチ（COMP-1/COMP-2到達時）にFIXMEコメントを追加

# テスト

`tests/run.src/extensions.at` に以下の11テストケースを追加した。

| テストケース | 内容 |
|---|---|
| CALL BY VALUE with BINARY | BINARY項目の基本的なBY VALUE渡し |
| CALL BY VALUE SIZE AUTO | SIZE AUTO指定 |
| CALL BY VALUE SIZE 8 | SIZE 8（8バイト/long）指定 |
| CALL BY VALUE UNSIGNED SIZE | UNSIGNED SIZE 4指定 |
| CALL BY VALUE with DISPLAY and PACKED-DECIMAL | DISPLAY項目とPACKED-DECIMAL項目の混在 |
| CALL BY VALUE with signed field | 符号付き項目（負の値） |
| CALL BY VALUE SIZE AUTO with DISPLAY large digits | SIZE AUTOで桁数が多い場合（8バイトへの昇格） |
| CALL BY VALUE with COMP-5 | COMP-5項目 |
| CALL BY VALUE with COMP-X | COMP-X項目 |
| CALL BY VALUE SIZE 1 and SIZE 2 | SIZE 1とSIZE 2の組み合わせ |
| CALL BY VALUE with literal | リテラル値のBY VALUE渡し |

# その他

- COMP-1/COMP-2（浮動小数点型）使用時の`default`ブランチは、現在COMP-1/COMP-2が未実装のため到達しない。COMP-1/COMP-2の実装時に合わせて対応する予定（FIXMEコメントに記載済み）。

### CODE REVIEW

File: cobj/codegen.c [hunk 1]
```diff
@@ -3155,33 +3150,19 @@ static void joutput_call(struct cb_call *p) {
...
-              joutput("short");
-              break;
-            case CB_SIZE_4:
-              joutput("int");
-              break;
             case CB_SIZE_8:
-              joutput("long long");
+              joutput("CobolDataStorage.primitiveToDataStorage((long)(");
+              joutput_integer(x);
+              joutput("))");
```
[the remark this answers is not in the document]
Comment (tsh-hashimoto) [reply cobj/codegen.c:3173]: Thank you for the review.

In this implementation, `primitiveToDataStorage(int)` always creates a 4-byte `CobolDataStorage`, so SIZE 1, 2, and 4 are all handled through the same `(int)` cast path. The callee interprets the received `CobolDataStorage` based on its own LINKAGE SECTION definition, so the byte size difference does not cause issues in practice.

Regarding the unsigned sign-extension concern: `joutput_integer(x)` generates Java code that retrieves the numeric value from the COBOL data item, and the `(int)` cast in Java handles the conversion correctly for the value ranges supported by COBOL `PIC 9` definitions (e.g., `PIC 9(2)` max is 99 for SIZE 1, `PIC 9(4)` max is 9999 for SIZE 2), which are well within the signed int range.

The existing test cases for SIZE 1 and SIZE 2 pass correctly with this approach.

File: cobj/codegen.c [hunk 2]
```diff
@@ -3194,6 +3175,9 @@ static void joutput_call(struct cb_call *p) {
             joutput(")");
             break;
           default:
+            /* FIXME: COMP-1 / COMP-2を使うとここを通ることがある
+             * COMP-1 / COMP-2を実装するときにここも見直す必要がある
+             */
             joutput("*(");
             joutput_data(x);
```
[the remark this answers is not in the document]
Comment (tsh-hashimoto) [reply cobj/codegen.c:3197]: Good catch — `*(...)` is indeed C pointer dereference syntax and not valid Java. However, as noted in the FIXME comment added in this PR, this `default` branch is only reachable when COMP-1 (float) / COMP-2 (double) types are used, which are not yet implemented in opensourcecobol4j. This means the branch is currently unreachable in practice.

The FIXME comment was intentionally added to flag this for future work when COMP-1/COMP-2 support is implemented. Addressing it is outside the scope of this PR.

File: tests/run.src/extensions.at [hunk 3]
```diff
@@ -1708,3 +1708,362 @@ prog.cob:9: Warning: Unreachable statement 'DISPLAY'
...
+       PROGRAM-ID.      caller.
+       DATA             DIVISION.
+       WORKING-STORAGE  SECTION.
+       01 WS-NUM1       PIC 9(2) BINARY VALUE 12.
+       01 WS-NUM2       PIC 9(4) BINARY VALUE 1234.
+       PROCEDURE        DIVISION.
+           CALL "callee" USING
+               BY VALUE SIZE 1 WS-NUM1
+               BY VALUE SIZE 2 WS-NUM2
+           END-CALL.
```
[the remark this answers is not in the document]
Comment (tsh-hashimoto) [reply tests/run.src/extensions.at:2124]: Thank you for the suggestion. The direction of testing edge cases is reasonable.

However, it is worth noting that COBOL PIC clauses constrain the value ranges: `PIC 9(2)` allows values 0–99 (SIZE 1) and `PIC 9(4)` allows values 0–9999 (SIZE 2). Values like 200 for SIZE 1 or 40000 for SIZE 2 would exceed the PIC definition and are not valid COBOL values for these fields. So the sign-extension issue with top-bit values does not arise in practice for correctly defined COBOL programs.

That said, adding unsigned SIZE 1/2 test cases could still be useful for completeness.

**Comment by tsh-hashimoto** [comment]
### Code review

Found 1 issue:

1. PR description lists 11 test cases (including "CALL BY VALUE with literal" / リテラル値のBY VALUE渡し), but the diff only adds 10 test cases. The literal test case is missing from `tests/run.src/extensions.at`. This leaves the `CB_TAG_LITERAL` code path in `cobj/codegen.c` untested.

https://github.com/opensourcecobol/opensourcecobol4j/blob/69c3dc18eb1ab0352aedf0314906b25425387243/cobj/codegen.c#L3091-L3100

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Comment by tsh-hashimoto** [comment]
テストケースを追加

| テストケース名 | 概要 | CALL文 | 期待値 | 対象コードパス |                                                                                                                                                                        
|---|---|---|---|---|                                                                                                                                                                                                               
| CALL BY VALUE with literal | 数値リテラルをBY VALUEで渡す | `BY VALUE 12345` / `BY VALUE 67890` | `000012345` / `000067890` | `codegen.c:3093-3095` (`CB_TAG_LITERAL` + `CB_CLASS_NUMERIC`) |                                     
| CALL BY VALUE with non-numeric literal | 非数値リテラル`"A"`をBY VALUEで渡す | `BY VALUE "A"` | `000000065` (ASCII 65) | `codegen.c:3097-3098` (`CB_TAG_LITERAL` + 非数値) |

File: cobj/codegen.c [hunk 4]
```diff
@@ -3155,33 +3150,19 @@ static void joutput_call(struct cb_call *p) {
...
-                joutput("(");
-              }
             }
             switch (sizes) {
-            case CB_SIZE_1:
-              joutput("char");
-              break;
-            case CB_SIZE_2:
-              joutput("short");
-              break;
```
Comment (yutaro-sakamoto) [inline cobj/codegen.c:3171]: CobolDataStorage.primitiveToDataStorage(short), CobolDataStorage.primitiveToDataStorage(byte)を新たに実装して、これらのメソッドをサイズが2や1の時にも使用するような実装にでしょうか?

**Comment by tsh-hashimoto** [comment]
サイズが1,2の場合でも、そのサイズに応じて`primitiveToDataStorage()`に変換されるように修正した。
それに伴いテストを追加した。以下が最終的なテスト内容。

| テストケース | 内容 |
|---|---|
| CALL BY VALUE with BINARY | BINARY項目の基本的なBY VALUE渡し |
| CALL BY VALUE SIZE AUTO | SIZE AUTO指定 |
| CALL BY VALUE SIZE 8 | SIZE 8（8バイト/long）指定 |
| CALL BY VALUE UNSIGNED SIZE | UNSIGNED SIZE 4指定 |
| CALL BY VALUE with DISPLAY and PACKED-DECIMAL | DISPLAY項目とPACKED-DECIMAL項目の混在 |
| CALL BY VALUE with signed field | 符号付き項目（負の値） |
| CALL BY VALUE SIZE AUTO with DISPLAY large digits | SIZE AUTOで桁数が多い場合（8バイトへの昇格） |
| CALL BY VALUE with COMP-5 | COMP-5項目 |
| CALL BY VALUE with COMP-X | COMP-X項目 |
| CALL BY VALUE SIZE 1 | SIZE 1（1バイト/byte）指定 |
| CALL BY VALUE SIZE 2 | SIZE 2（2バイト/short）指定 |
| CALL BY VALUE SIZE 4 | SIZE 4（4バイト/int）指定 |
| CALL BY VALUE with literal | リテラル値のBY VALUE渡し |
| CALL BY VALUE with non-numeric literal | 非数値リテラルのBY VALUE渡し |

File: cobj/codegen.c (hunk 4 above)
Comment (tsh-hashimoto) [reply cobj/codegen.c:3171]: サイズ1,2の場合でもそれぞれ別々の`primitiveToDataStorage()`に変換されるようにして、`short`や`int`も渡せるように新しく関数を実装しました。
テストも新しく追加しています。

**Review by yutaro-sakamoto — APPROVED** [review]
